wled00/FX_fcn.cpp

```diff
@@ -637,37 +637,31 @@ uint16_t IRAM_ATTR_YN Segment::nrOfVStrips() const {
-constexpr int Pinwheel_Size_Small  = 16;       // larger than this -> use "Medium"
-constexpr int Pinwheel_Steps_Medium = 192;     // no holes up to 32x32
-constexpr int Pinwheel_Size_Medium  = 32;      // larger than this -> use "Big"
-constexpr int Pinwheel_Steps_Big = 304;        // no holes up to 50x50
-constexpr int Pinwheel_Size_Big  = 50;         // larger than this -> use "XL"
-constexpr int Pinwheel_Steps_XL  = 368;
-constexpr float Int_to_Rad_Small = (DEG_TO_RAD * 360) / Pinwheel_Steps_Small;  // conversion: from 0...72 to Radians
-constexpr float Int_to_Rad_Med =   (DEG_TO_RAD * 360) / Pinwheel_Steps_Medium; // conversion: from 0...192 to Radians
-constexpr float Int_to_Rad_Big =   (DEG_TO_RAD * 360) / Pinwheel_Steps_Big;    // conversion: from 0...304 to Radians
-constexpr float Int_to_Rad_XL =    (DEG_TO_RAD * 360) / Pinwheel_Steps_XL;     // conversion: from 0...368 to Radians
-
-constexpr int Fixed_Scale = 512;               // fixpoint scaling factor (9bit for fraction)
+constexpr int Fixed_Shift = 15; // fixpoint scaling factor (15bit for fraction)
 
-// Pinwheel helper function: pixel index to radians
-static float getPinwheelAngle(int i, int vW, int vH) {
-  int maxXY = max(vW, vH);
-  if (maxXY <= Pinwheel_Size_Small)  return float(i) * Int_to_Rad_Small;
-  if (maxXY <= Pinwheel_Size_Medium) return float(i) * Int_to_Rad_Med;
-  if (maxXY <= Pinwheel_Size_Big)    return float(i) * Int_to_Rad_Big;
-  // else
-  return float(i) * Int_to_Rad_XL;
-}
 // Pinwheel helper function: matrix dimensions to number of rays
-static int getPinwheelLength(int vW, int vH) {
+static inline int getPinwheelSteps(int vW, int vH) {
   int maxXY = max(vW, vH);
-  if (maxXY <= Pinwheel_Size_Small)  return Pinwheel_Steps_Small;
-  if (maxXY <= Pinwheel_Size_Medium) return Pinwheel_Steps_Medium;
-  if (maxXY <= Pinwheel_Size_Big)    return Pinwheel_Steps_Big;
-  // else
-  return Pinwheel_Steps_XL;
+  return maxXY * 5; // theoretical value sqrt(2) * pi (1 pixel step in the corner) Note: (maxXY * 9)/2 works on most sizes but not all, there is room for speed optimization here
+}
+// Pinwheel helper function: pixel index to radians
+static float getPinwheelAngle(int i, int vW, int vH) {
+  int steps = getPinwheelSteps(vW, vH) - 1; // -1 to make the angle larger so the wheel has no gap
+  return float(i) * (DEG_TO_RAD * 360) / steps;
+}
+
+static void setPinwheelParameters(int i, int vW, int vH, unsigned& posx, unsigned& posy, int& inc_x, int& inc_y, int& maxX, int& maxY) {
+  float angleRad = getPinwheelAngle(i, vW, vH); // angle in radians
+  float cosVal = cos_t(angleRad);
+  float sinVal = sin_t(angleRad);
+  posx = (vW - 1) << (Fixed_Shift - 1); // X starting position = center (in fixed point)
+  posy = (vH - 1) << (Fixed_Shift - 1); // Y starting position = center (in fixed point)
+  inc_x = cosVal * (1 << Fixed_Shift);  // X increment per step (fixed point)
+  inc_y = sinVal * (1 << Fixed_Shift);  // Y increment per step (fixed point)
```

Comment:
Same comment as below - generally it's better to avoid shifting signed integers.
And "1 << bits" is harder to read than "32768" which is a well-known power of two.

I would propose to stay with my original implementation, using the scaling factor instead shifting. Values around.

You could even translate one into the other: `constexpr unsigned fixed_scale = 1U << fixed_shift;`

Reply:
Update: this page from Microsoft explains the problem: https://learn.microsoft.com/en-us/cpp/cpp/left-shift-and-right-shift-operators-input-and-output?view=msvc-170

I think that c++20 (the c++ language definition from 2020) finally ended this by stating "Right-shift on signed integral types is an arithmetic right shift, which performs sign-extension.". But for older compilers, I would still not rely on "what happens when I shift  negative value".

Some more explanation is here: https://stackoverflow.com/questions/7522346/right-shift-and-signed-integer

Reply:
I read shifts easier than large numbers.

Reply:
Then maybe `constexpr int fixed_scale = 1U << fixed_shift;. ` is a compromise that keeps both of us happy 😉 sorry again for me going into "lecture mode".